packages/core-utils/src/nav/nav-utils.ts

```diff
@@ -0,0 +1,27 @@
+interface KeyedNavFn<Navs, FnReturn> {
+  <K extends keyof Navs>(key: K, args: Navs[K]): FnReturn;
+}
+
+/**
+ * Provides type safety for navigation mechanics.
+ * navTo and getUrl (key, args) functions are typed based on NavTypes generic to
+ * narrow args based on the type of NavTypes[key] contract.  NavTypes can hold a
+ * number of keyed arg contracts.
+ *
+ * @example
+ * interface MyNavAction {
+ *   key1: { a: string, b: number },
+ *   key2: null
+ * }
+ *
+ * // using NavLinkProvider<MyNavAction> type will have getUrl narrow
+ * // its 2nd argument based on the first argument:
+ *
+ * myProvider.getUrl('key1', args: {a: string, b: number}) => string
+ *
+ * // and similar for navTo
+ */
+export interface NavLinkProvider<NavTypes> {
+  navTo: KeyedNavFn<NavTypes, void>;
```

Comment:
I don't quite see why `navTo` is part of this interface. It seems to me that a generic `goToUrl(string)` would be enough.

Reply:
We'd want components to declare what routes they might attempt to navigate to so that consumers can ensure that their application can handle those navigation requests. I think a generic `goToUrl(string)` would require a consumer to read the source code to understand all URLs that the component might try to navigate to.

Also, URL instead of route name means the application's route names becomes bound by the component's. Theoretically, an application should be able to name the route `/envs` even if the component calls it 'environments'.

Reply:
As we discussed, I can add some thinking on this to the interface here.   NavTo was chosen to fully remove the "concern" of navigation and any side-effect of that that different nav solutions may need to handle.

Reply:
documentation tweaked

Reply:
I get why we renamed `getLink` to `getUrl` (it's a more descriptive name).

But does the name of this method need to diverge so much from existing code? I would prefer something like `goToRoute`: the `go` to mirror the names in Terra UI nav and the [History](https://developer.mozilla.org/en-US/docs/Web/API/History) API, and `route` to clarify that this doesn't take a URL as one might assume for a navigation API.

Reply:
my thinking is even route is an implementation detail... the generic concern is "navigation"...